client/Fireport/Assets/Scripts/AI/DumbAI.cs

```diff
@@ -0,0 +1,103 @@
+                return _api;
+            }
+        }
+
+        void Start() {
+            DontDestroyOnLoad(gameObject);
+            WebSocketListener.Instance().Subscribe(this);
+        }
+        
+        private void RefreshApi() {
+            _api = FindObjectOfType<RestApi>();
+            if (!_api) throw new Exception("A RestApi object is required in the scene for a DumpAI");
+        }
+
+        public void Initialize() {
+            gameCode = LobbyInfoController.Instance().msg.id;
+            playerName = string.Format("DumbAI_{0}", Ais.Count);
+            gameObject.name = playerName;
+            Ais.Add(this);
+            StartCoroutine(WaitThenDo(1, () => JoinLobby()));
+        }
+
+        public void JoinLobby() {
+            Api.JoinLobby(gameCode, playerName, s => {
+                Debug.Log(string.Format("{0} just joined the lobby: {1}", playerName, s));
+                Api.ReadyUp(gameCode, playerName, () => {
+                    Debug.Log(string.Format("{0} just readied up", playerName));
+                });
+            });
+        }
+
+        public void PlayCards() {
+            Api.SubmitTurn(gameCode, gameState.Turn, playerName, playerId, GetCardsFromHand(3).ConvertAll(c => c.ID).ToArray(), () => Debug.Log(string.Format("{0} just played their cards", playerName)));
+        }
+
+        public List<CardState> GetCardsFromHand(int count) {
+            var player = gameState.Players.Find(p => p.ID == playerId);
+            List<CardState> cards = new List<CardState>();
+            for (int i = 0; i < player.Hand.Count && i < count; i++) {
+                cards.Add(player.Hand[i]);
+            }
+
+            return cards;
+        }
+
+        public void handleDownStreamMessage(string messageType, string message) {
+            switch (messageType) {
+                case MsgTypes.GAME_START:
+                    var gameStartMessage = JsonUtility.FromJson<GameStartMessage>(message);
+                    gameState = gameStartMessage.gameState;
+                    playerId = gameState.Players.Find(p => p.Name == playerName).ID;
+                    PlayCards();
+                    break;
+                case MsgTypes.TURN_RESULT:
+                    var turnResultMessage = JsonUtility.FromJson<TurnResultMessage>(message);
+                    gameState = turnResultMessage.currentState;
+                    playerId = gameState.Players.Find(p => p.Name == playerName).ID;
+                    PlayCards();
+                    break;
+                default:
+                    Debug.Log(string.Format("{0} failed to handle message type({1}): {2}", playerName, messageType, message));
+                    break;
+            }
+        }
+
+        private void OnDestroy() {
+            WebSocketListener.Instance().Unsubscribe(this);
+        }
+
+        private IEnumerator WaitThenDo(float secondsToWait, Action thenDo) {
+            yield return new WaitForSeconds(secondsToWait);
```

Comment:
dafuq does `yield` do?

Reply:
It is C#'s stupid way of handling some asynchronous stuff.  To be perfectly honest, I don't really *get* it, I just kind of know when I'm suppose to use it in Unity.  To me, it doesn't feel very intuitive at all.  But that is usually the case when there is a feature in a language that doesn't map to any other language you already know.